ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/QueryEndpoint.java

```diff
@@ -144,26 +149,33 @@ private QueryPublisher createPullQueryPublisher(
     );
 
     PullQueryExecutionUtil.checkRateLimit(rateLimiter);
+    final Decrementer decrementer = pullConcurrencyLimiter.increment();
+
+    try {
+      final PullQueryResult result = ksqlEngine.executePullQuery(
+          serviceContext,
+          statement,
+          routing,
+          routingOptions,
+          plannerOptions,
+          pullQueryMetrics,
+          false
+      );
+
+      result.onCompletionOrException((v, t) -> decrementer.decrementAtMostOnce());
```

Comment:
nit: we can simplify to register only one callback, which checks if `t == null` or not and determine whether to record latency.

Reply:
Sure, makes sense.